src/main/java/edu/cornell/kfs/sys/service/impl/AwsSecretServiceImpl.java

```diff
@@ -24,26 +28,62 @@
 
 import edu.cornell.kfs.sys.CUKFSConstants;
 import edu.cornell.kfs.sys.service.AwsSecretService;
+import edu.cornell.kfs.sys.util.CallableForThrowType;
+import edu.cornell.kfs.sys.util.CuJsonUtils;
 
 public class AwsSecretServiceImpl implements AwsSecretService {
     private static final Logger LOG = LogManager.getLogger(AwsSecretServiceImpl.class);
-    
+
+    private static final ThreadLocal<LinkedList<Map<String, String>>> AWS_SECRETS_CACHE_STACK =
+            ThreadLocal.withInitial(LinkedList::new);
+
     protected String awsRegion;
     protected String kfsInstanceNamespace;
     protected String kfsSharedNamespace;
     protected int retryCount;
     
+    @Override
+    public <R, T extends Throwable> R doWithAwsSecretsCachingEnabled(
+            CallableForThrowType<R, T> callable) throws T {
+        Map<String, String> awsSecretsCache = new HashMap<>();
+        boolean pushedCacheToStack = false;
+        try {
+            getAwsSecretsCacheStack().offerFirst(awsSecretsCache);
+            pushedCacheToStack = true;
+            return callable.call();
+        } finally {
+            try {
+                awsSecretsCache.clear();
+                if (pushedCacheToStack) {
+                    getAwsSecretsCacheStack().pollFirst();
+                }
+            } catch (RuntimeException e) {
+                LOG.error("doWithAwsSecretsCachingEnabled: Unexpected error when clearing and removing cache", e);
```

Comment:
I had concerns about the Callable's thrown exception (if any) getting lost if this nested try-catch had its error rethrown, but I agree that it would be good to rethrow this nested block's exception if it did end up with an unexpected problem. I'll rework this to add better error handling and error logging, so that we can keep better track of any exceptions thrown by the Callable and the nested cleanup block.